include/irr/asset/filters/CSummedAreaTableImageFilter.h

```diff
@@ -152,145 +172,183 @@ class CSummedAreaTableImageFilter : public CMatchedSizeInOutImageFilterCommon, p
-						*(currentColumnCache + channel) += *(mainRowCacheScratch + channel) += decodeBuffer[channel];
-						*(finalPixel + channel) = *(currentColumnCache + channel) - (ExclusiveMode ? decodeBuffer[channel] : 0);
-					}
+						const auto offset = ((readBlockPos.z * state->extent.height + readBlockPos.y) * state->extent.width + readBlockPos.x) * currentChannelCount;
 
-					memcpy(scratchMemory + outDataAdressOffsetScratch + layersOffset, finalPixel, state->mainRowScratchCacheByteSize);
+						if (readBlockPos.w == w)
+							return scratchMemory + offset;
+						else
+							return reinterpret_cast<decodeType*>(stateCopy.scratchMemory) + offset;
+					};
 
-					if (newReadBlockPos.x == trueOutExtent.width - 1) // reset row cache when changing y
+					auto sum = [&](core::vectorSIMDi32 readBlockPos) -> void
 					{
-						if (newReadBlockPos.y == 0) // sum values in (maxX, 0, z) and remove values from column cache
-						{ 
-							memcpy(totalImageCacheScratch + (newReadBlockPos.w * state->extent.depth + newReadBlockPos.z) * state->mainRowScratchCacheByteSize, finalPixel, state->mainRowScratchCacheByteSize);
-							memset(mainColumnCacheScratch, 0, state->mainColumnScratchCacheByteSize);
+						auto current = getScratchPixel(readBlockPos);
+
+						auto addScratchPixelToCurrentOne = [&](decodeType* values)
+						{
+							for (auto i = 0; i < currentChannelCount; ++i)
+								current[i] += values[i];
+						};
+						
+						auto substractScratchPixelFromCurrentOne = [&](decodeType* values)
+						{
+							for (auto i = 0; i < currentChannelCount; ++i)
+								current[i] -= values[i];
+						};
+
+						addScratchPixelToCurrentOne(getScratchPixel(readBlockPos - core::vectorSIMDi32(0, 0, 1, 0)));					// add box x<=current_x && y<=current_y && z<current_z
+						addScratchPixelToCurrentOne(getScratchPixel(readBlockPos - core::vectorSIMDi32(0, 1, 0, 0)));					// add box x<=current_x && y<current_y && z<=current_z
+						substractScratchPixelFromCurrentOne(getScratchPixel(readBlockPos - core::vectorSIMDi32(0, 1, 1, 0)));		    // remove overlap box x<=current_x && y<current_y && z<current_z
+						
+						// now I have the sum of all layers below me in the bound x<=current_x && y<=current_y && z<current_z  and the current pixel
+						// time to add the missing top layer pixels
+
+						addScratchPixelToCurrentOne(getScratchPixel(readBlockPos - core::vectorSIMDi32(0, 1, 0, 0)));					// add box x<current_x && y<=current_y && z<=current_z
+						substractScratchPixelFromCurrentOne(getScratchPixel(readBlockPos - core::vectorSIMDi32(0, 1, 0, 1)));			// remove overlap box x<current_x && y<=current_y && z<current_z
+						substractScratchPixelFromCurrentOne(getScratchPixel(readBlockPos - core::vectorSIMDi32(0, 1, 1, 0)));			// remove future overlap box x<current_x && y<current_y && z<=current_z
+						addScratchPixelToCurrentOne(getScratchPixel(readBlockPos - core::vectorSIMDi32(0, 1, 1, 1)));					// add box x<current_x && y<current_y && z<current_z
```

Comment:
see my new code listing in the comments, I had the offsets wrong for the second part of the sum (the last 4 lines here)

Also there should be no addition/subtraction if the readBlockPos coordinate is 0 in the component(s) that will get offset by `-1` [also mentioned in my revised listing]

Reply:
corrected